qiskit/utils/run_circuits.py

```diff
@@ -381,3 +381,184 @@ def run_on_backend(backend: Union[Backend, BaseBackend],
     else:
         job = backend.run(qobj, **backend_options, **noise_config)
         return job
+
+
+def run_circuits(circuits: Union[QuantumCircuit, List[QuantumCircuit]],
+                 backend: Union[Backend, BaseBackend],
+                 qjob_config: Dict,
+                 backend_options: Optional[Dict] = None,
+                 noise_config: Optional[Dict] = None,
+                 run_config: Optional[Dict] = None,
+                 job_callback: Optional[Callable] = None) -> Result:
+    """
+    An execution wrapper with Qiskit-Terra, with job auto recover capability.
+
+    The auto-recovery feature is only applied for non-simulator backend.
+    This wrapper will try to get the result no matter how long it takes.
+
+    Args:
+        circuits: circuits to execute
+        backend: backend instance
+        qjob_config: configuration for quantum job object
+        backend_options: backend options
+        noise_config: configuration for noise model
+        run_config: configuration for run
+        job_callback: callback used in querying info of the submitted job, and
+                                           providing the following arguments:
+                                            job_id, job_status, queue_position, job
+
+    Returns:
+        Result object
+
+    Raises:
+        QiskitError: Any error except for JobError raised by Qiskit Terra
+    """
+    backend_options = backend_options or {}
+    noise_config = noise_config or {}
+    run_config = run_config or {}
+    with_autorecover = not is_simulator_backend(backend)
+
+    job, job_id = _safe_submit_circuits(circuits,
+                                        backend,
+                                        qjob_config=qjob_config,
+                                        backend_options=backend_options,
+                                        noise_config=noise_config,
+                                        run_config=run_config)
+    result = None
+    if with_autorecover:
+        logger.info("Backend status: %s", backend.status())
+        logger.info("There is one jobs are submitted: id: %s", job_id)
+        while True:
+            logger.info("Running job id: %s", job_id)
+            # try to get result if possible
+            while True:
+                job_status = _safe_get_job_status(job, job_id)
+                queue_position = 0
+                if job_status in JOB_FINAL_STATES:
+                    # do callback again after the job is in the final states
+                    if job_callback is not None:
+                        job_callback(job_id, job_status, queue_position, job)
+                    break
+                if job_status == JobStatus.QUEUED:
+                    queue_position = job.queue_position()
+                    logger.info("Job id: %s is queued at position %s", job_id, queue_position)
```

Comment:
This isn't actually part of the jobv1 interface:

https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/providers/job.py#L35-L141

afaik the only job type that has this method is ibmq's. Calling this unconditionally will cause trouble for any providers that have jobs in a queued state I think maybe just make this opportunistically called something like:

```suggestion
                if job_status == JobStatus.QUEUED and hasattr(job, queue_position):
                    queue_position = job.queue_position()
                    logger.info("Job id: %s is queued at position %s", job_id, queue_position)
```